crates/bevy_ecs/src/system/function_system.rs

```diff
@@ -77,6 +79,8 @@ impl SystemMeta {
 /// - [`Local`](crate::system::Local) variables that hold state
 /// - [`EventReader`](crate::event::EventReader) system parameters, which rely on a [`Local`](crate::system::Local) to track which events have been seen
 ///
+/// Note that this is automatically handled for you when using a [`SystemRegistry`](crate::system::SystemRegistry).
```

Comment:
Thinking about this I don't think this would be the desired behavior for most use cases. Consider that UI callbacks will often be the system that just operate on different entity patterns in their local hierarchy. The state for these should not be shared. Like if I have buttons that increment a counters and I use `Local`s to store the state of those counters those buttons should have their own individual state and anything shared should be done through resources. Same for event readers and other system params. 

I understand that currently systems can only be registered once. This should be changed to allow for multiple registrations in effect allowing for multiple local states.

Reply:
Yeah, I think that's a fair design critique. Being forced to use closures to get the unique identifiers would be a pain here.

Reply:
I understand why you'd want to have the feature, and from an implementation standpoint it isn't that bad, but the mental overhead it creates is very annoying to deal with.

Take the following example:
```rust
world.run_system(my_system);
world.run_system(my_system);
```
This system should run twice, and if it contains `Local`s then the state of those `Local`s is maintained between runs.

This makes sense.

Let's now do the following:
```rust
world.run_system(my_system); // register one system & run it
let system_id = SystemRegistry::register(my_system); // register the system again
world.run_system(my_system); // what should run ?
```
Now it's a real question of what system should run on the third line. Which of the two systems should be run again, or should a system be registered for the third time?
The most infuriating thing about this is that this problem only comes into play in the context of `Local`s. For any other context, it's perfectly fine to have the mental model of a 1-to-1 function name-to-system correspondence. 

I think that for now, it's reasonable to keep a simple API. If people want to run a system they can run a system, without having to worry about what they're running exactly.

Reply:
The underlying assumption I'm making is that it's by far more desirable to call systems by their name than having to keep track of system ids. An API that would consist solely of `SystemRegister::register(my_system)` and `SystemRegister::run_system_by_id(system_id)`, would be perfectly sound and consistent.

Reply:
> The underlying assumption I'm making is that it's by far more desirable to call systems by their name
What would this be used for tho?

Reply:
"Look this thing up by its ID/label" is very common in our API, so I don't think it'd look out-of-place here, but I'm OK to approve without it and let user feedback guide us.

If we store systems as entities in the future, users could have as many instances of `foo` as they want. The only thing that would have to be unique is the type/name -> entity mapping in some index.

Reply:
> > The underlying assumption I'm making is that it's by far more desirable to call systems by their name
> 
> What would this be used for tho?

It can be used for debugging systems, more complex control flow then is currently possible within a schedule and refactoring large systems into smaller ones:

```rust
if  checkmate {
    commands.run_system(end_game);
} else {
    commands.run_system(ai_make_move);
}
```

Reply:
Not sure what you mean by debugging systems but that control flow example is something I would be against. Wrong tool for the job.

Reply:
It's now possible to register a system as many times as you want.